Integration tests for GTalk

dborkan opened the pull request:
Adding 3 integration tests for GTalk (running in a Chrome app).  Testing that we:
* Can login and logout
* Can send a message from Alice to Bob
* Can send 8 messages from Alice to Bob in 1 second without throttling (503 "service unavailable" errors)

We've reached out to our contact on GTalk XMPP to learn more about the specifics about GTalk throttling.  For now sending 8 message per second (separated by 110ms) consistently works.  We will update these tests and possibly the socialprovider.js code as we learn more about the specific behavior of GTalk.

Also contains fixes for https://github.com/freedomjs/freedom-social-xmpp/issues/82

ryscheng commented on package.json:
```diff
@@ -20,12 +20,16 @@
     "prepublish": "grunt test"
   },
   "devDependencies": {
+    "async": "^0.9.0",
```
What is this for? Similarly for all new dependencies:
async, chalk, fs-extra, glob

dborkan replied:
These are all required by freedom-for-chrome's integration task (https://github.com/freedomjs/freedom-for-chrome/blob/master/tasks/integration.js)

ryscheng replied:
yuck. Okay, filed 2 new issues
https://github.com/freedomjs/freedom-for-firefox
https://github.com/freedomjs/freedom-for-chrome/issues/58

I believe @willscott is taking care of these today and hopefully we can get rid of these dependencies here

ryscheng replied:
See https://www.npmjs.com/package/grunt-jasmine-chromeapp

ryscheng replied:
I believe we can get rid of these dependencies now.

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+
+// GTalk gives alicefreedomxpp an anoymized id for bobfreedomxmpp
+// and vice-versa.
+var ALICE_ANONYMIZED_ID = '3bbk93gplwqcn3abf19ivuh6s5@public.talk.google.com';
```
Are these stable? Or do they change over time?

dborkan replied:
AFAIK they are stable..  Although this test will alert us if they are not.

ryscheng replied:
Hm. Trying to think if there's anyway to do this that doesn't require the anonymized ID. Is it unsafe to assume that there are only 2 people on their respective buddy lists? (themselves and the other)

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+    });
+  });
+
+  it('Can send messages', function(done) {
```
Should we add some expect statements to check the values?

dborkan replied:
I think the 'Can send messages' and 'Can send 8 messages per second' test cases are fine without expect statements.  Adding an expect that the messageData.message == uniqueMsg is redundant to the check we already have - and I think its better not to add that expect statement for all messages since there is a small chance bobSocialClient will receive a message from somewhere outside this test

ryscheng replied:
This might be silly, but maybe we should send unique messages every time and make sure that the TOTAL_MESSAGES that we're getting are the unique ones we expect (as opposed to duplicates)

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+  // We should be able to send 8 messages per second from one peer to another
+  // without being throttled by GTalk (i.e. we should not get 503 "service
+  // unavailable" errors).
+  it('Can send 8 messages per second', function(done) {
```
Similarly, let's add some expect statements

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+              messageData.message == uniqueMsg) {
+            // Keep this trace so we know how many messages are received
+            // in case of failure.
+            console.log('received message');
```
Replace log with expects?

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+    // Defined in socialprovider.js
+    // TODO: make this a constant available to users of the freedom social
+    // provider object
+    var BATCH_INTERVAL_MS = 100;
```
What's the meaning of this? It's never used?

dborkan replied:
Just used to calculate MESSAGE_FREQUENCY.  I could remove it and just do ```var MESSAGE_FREQUENCY = 110;``` but I thought having this extra constant is more clear

dborkan replied:
Just made that change with a comment explaining how 110 is calculated

ryscheng replied:
Oh okay

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+  // without being throttled by GTalk (i.e. we should not get 503 "service
+  // unavailable" errors).
+  it('Can send 8 messages per second', function(done) {
+    var NUM_MESSAGES_PER_SECOND = 8;
```
I think this is just total messages sent right?

dborkan replied:
Yes

dborkan replied:
Just renamed to TOTAL_MESSAGES

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+            // Keep this trace so we know how many messages are received
+            // in case of failure.
+            console.log('received message');
+            if (++receivedMessageCount == NUM_MESSAGES_PER_SECOND) {
```
Not a fan of inline incrementing. Can we make this an explicit statement?

dborkan replied:
Done

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+  // is running the same tests at the same time.
+  var uniqueMsg = Math.random().toString();
+
+  function logoutThenDone(socialClientsArray, done) {
```
Some of these helper functions are outside the describe. Can we put them all in 1 place?
Maybe a var Helper = {waitForUser: ..., logoutThenDone: ...} etc?

dborkan replied:
Done

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+  return getAccessToken(refreshToken).then(function(accessToken) {
+    // Set the accessToken to pendingAccessToken so that it can be used by
+    // our OAuthView.
+    pendingAccessToken = accessToken;
```
This seems brittle to have a global pending access token.
What if we got both access tokens, create 2 OAuthProvider functions with these tokens bound, and used that in the beforeEach to instantiate freedom? Or maybe something else. Thinking about how to clean this up a bit.

dborkan replied:
I agree having the global isn't great, but I'm not sure I understand the proposed solution.  The problem is that there is no way to pass an argument to socialClient.login that will then make its way to OAuthProvider.launchAuthFlow.  If there was then I could pass an email address or just the accessToken and not have any global state.

ryscheng replied:
My suggestion is to bind it to the oAuth class on Line 148, basically call freedom twice and create 2 socialinterfaces

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+        '&client_id=' + CLIENT_ID +
+        '&client_secret=h_hfPI4jvs9fgOgPweSBKnMu' +
+        '&grant_type=refresh_token';
+    ajax('POST', 'https://www.googleapis.com/oauth2/v3/token', data)
```
Seems like ajax is only used once, maybe just inline it here?

ryscheng commented:
Super cool. really glad we have this now! just some small cleanup comments

dborkan commented:
Just made most of the small changes.  I didn't add any new expect statements because they seem redundant to the if checks to verify that messageData.message==uniqueMsg.  Also didn't make any changes yet to the OAuthProvider..  PTAL, thanks

ryscheng commented on Gruntfile.js:
```diff
@@ -135,6 +135,25 @@ module.exports = function(grunt) {
+          helper: [
+            // "include: true" is needed for dependencies that are used by
+            // the jasmine tests in the core environment.
+            {path: 'dist/social.google.json', include: false},
```
This should use glob notion.

dborkan replied:
What notation are you thinking?  I don't think we want to include everything in dist/, for example both google-auth.js and facebook-auth.js will clobber each other.

ryscheng replied:
I think we do want dist/*.js right? Why would they clobber each other? It's up to the manifest file to choose which one it wants. Thats why we also have 3 manifests, social.google.json, social.facebook.json, and social.xmpp.json

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,237 @@
+/*
```
Very confused. I think this same file is checked in twice?
spec/integration/gtalk.spec.js
spec/chrome_integration/gtalk.spec.js

dborkan replied:
Sorry, I meant to move the file from spec/integration to
spec/chrome_integration but accidentally made a duplicate.  This is fixed
now

On Thu Jan 22 2015 at 4:01:27 PM Raymond Cheng <notifications@github.com>
wrote:

> In spec/chrome_integration/gtalk.spec.js
> <https://github.com/freedomjs/freedom-social-xmpp/pull/85#discussion_r23409342>
> :
>
> > @@ -0,0 +1,237 @@
> > +/*
>
> Very confused. I think this same file is checked in twice?
> spec/integration/gtalk.spec.js
> spec/chrome_integration/gtalk.spec.js
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/freedomjs/freedom-social-xmpp/pull/85/files#r23409342>
> .
>

dborkan commented:
Just made a ton of changes, please take a look again:
* OAuth no longer relies on a global pending access token.  Instead we setup 2 OAuth providers, each with a different refresh token.
* Credentials are now in gtalk_credentials.js with more instructions in both that file and gtalk.spec.js
* Now if Alice or Bob has multiple online clients, we send messages to all of them (rather than the first to appear in the onClientState event)
* We no longer need to hardcode the anonymized userId in JavaScript, and instead can get this by creating a map from the user name to the anonymized id

ryscheng commented on Gruntfile.js:
```diff
@@ -173,6 +193,7 @@ module.exports = function(grunt) {
   grunt.loadNpmTasks('grunt-prompt');
   grunt.loadNpmTasks('grunt-bump');
   grunt.loadNpmTasks('grunt-npm');
+  grunt.loadTasks('node_modules/freedom-for-chrome/tasks');
```
As of today, we have separated this out into a separate repo/npm module
https://www.npmjs.com/package/grunt-jasmine-chromeapp

Let's use that instead

dborkan replied:
Done

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,240 @@
+      // Next login as Bob.
+      var bobSocialClient = bobSocialInterface();
+      Helper.loginAs(bobSocialClient, BOB_EMAIL).then(function(bobClientInfo) {
+        bobSocialClient.on('onMessage', function(messageData) {
```
Shouldn't this callback be registered before Helper.loginAs(bobSocialClient)?

dborkan replied:
Done

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,240 @@
+      if (clientState.userId == userId &&
+          clientState.status == 'ONLINE' &&
+          !onlineClientIds[clientState.clientId]) {
+        // Mark this client as online so we don't re-invoke the callback
```
I'm a little bit worried here. Shouldn't this lead to bugs?
onClientOnline is used in multiple tests, so onlineClientIds is not cleared between tests right?
Is there a better way to do this?

dborkan replied:
For each test we create new socialClient objects, and will call onClientOnline again as needed, so onlineClientIds will effectively be reset between each test.

ryscheng replied:
Can you clarify? Do we expect clientIds for both Alice and Bob to be different between tests?

One potential solution:
We should create an afterEach(...) clause that does necessary cleanup after every single test.
This would handle logging out of each socialClient.
onlineClientIds can be a property off of Helper, which would be cleared in the afterEach as well.

What do you think about that?

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,240 @@
+  // Returns a promise that fulfills when socialClient has been logged in
+  // using the loginEmail address.  loginEmail must have an associated
+  // refresh token.
+  loginAs: function(socialClient, loginEmail) {
```
loginEmail not used?

ryscheng replied:
Is this helper function necessary?

dborkan replied:
Just removed

ryscheng commented on spec/chrome_integration/gtalk.spec.js:
```diff
@@ -0,0 +1,240 @@
+      var bobSocialClient = bobSocialInterface();
+      Helper.loginAs(bobSocialClient, BOB_EMAIL).then(function(bobClientInfo) {
+        var receivedMessageCount = 0;
+        bobSocialClient.on('onMessage', function(messageData) {
```
Shouldn't this onMessage handler be registered before login?

dborkan replied:
Done

ryscheng commented:
Looks much cleaner! Woo

dborkan commented:
Just made a bunch more changes, PTAL
* Moved to grunt-jasmine-chromeapp
* Removed loginAs method
* Changed format in gtalk_credentials to use an ALICE and BOB object (makes it clear which fields are needed and you can easily switch ALICE and BOB if you want).
* Moved all onMessage listeners to before login

I didn't make any changes to onClientOnline, although I agree its the not perfect..  I want something that encapsulates listening for a specific userId to sign online and invoking a callback each time a new clientId comes online for that userId - but doesn't generate extra callbacks when only timestamp fields have changed.

ryscheng commented:
One potential solution:
We should create an afterEach(...) clause that does necessary cleanup after every single test.
This would handle logging out of each socialClient.
onlineClientIds can be a property off of Helper, which would be cleared in the afterEach as well.

What do you think about that?

dborkan commented:
Cool, some updates:
* I just made those changes to add an afterEach that resets the onlineClientIds map and logs out Alice and Bob.
* Refactored beforeEach a bit, now beforeEach always sets aliceSocialClient and bobSocialClient to a new social client object (but doesn't call login).
* Removed an unused var in socialprovider.js (and verified that calling logout without login first is fine).

PTAL

ryscheng commented:
Awesome! This test is super slick. Thanks for putting this together!
*digihugs

:+1:
